analyzer/consensus/consensus.go

```diff
@@ -636,7 +743,7 @@ func (m *Main) queueMetadataRegistry(ctx context.Context, batch *storage.QueryBa
 	return nil
 }
 
-func (m *Main) prepareStakingData(ctx context.Context, height int64, batch *storage.QueryBatch) error {
+func (m *Main) prepareStakingData(ctx context.Context, height int64, batch *storage.QueryBatch, txHashIndexLookup map[string]int) error {
```

Comment:
There's gotta be a better way than pass this data all the way down right? 💭

Reply:
There's ... using the data that was inserted into SQL with previous statements? :)

I'll re-post this suggestion here to save it from Slack evisceration:
```sql
INSERT INTO oasis_3.events (backend, type, body, txn_block, txn_hash, txn_index)
SELECT ($1, $2, $3, $4, $5, txn_index) FROM oasis_3.transactions
WHERE txn_block=$4 AND txn_hash=$5 -- AND code=0 (code 0 means success; but a non-successful tx could also generate events)
UNION
SELECT ($1, $2, $3, $4, $5, -1)  -- dummy value (that will fail FK constraint)
ORDER BY txn_index DESC LIMIT 1; -- prefer non-dummy value if SELECT succeeds
```

If we use this version of the insert statement, we don't have to pass anything around in golang.

That said, I now see that we have multiple insert statements of this kind (= where we need the tx_idx). In which case, the code duplication above is not great (as we have modify multiple `INSERT` statements with the same heuristic code), and it'd be better to have a function. We'd define this in migrations:
```sql
-- Returns the likely index of transaction `_tx_hash` in block at `_height`.
CREATE FUNCTION guess_tx_index(_height BIGINT, _tx_hash TEXT) RETURNS INT AS $$
  SELECT txn_index FROM oasis_3.transactions
  WHERE txn_block=_height AND txn_hash=_tx_hash
  ORDER BY code ASC  -- code==0 means success, >0 is for failures; prefer successful txs
  LIMIT 1
$$ LANGUAGE SQL
```
then when inserting data, we'd call the function:
```sql
INSERT INTO oasis_3.events (backend, type, body, txn_block, txn_hash, txn_index)
VALUES ($1, $2, $3, $4, $5, COALESCE(guess_tx_index($4, $5), -1))
```
The `COALESCE` is optional and is just there to explicitly violate the FK with `-1` if `guess_tx_index` returns NULL -- that should never happen if input data is clean and our code is bug-free.

Reply:
Good record keep, but we no longer will pass this index data down. Whew.